### PR TITLE
Add support for seccomp thread sync feature

### LINKED ISSUE #57: Support for SECCOMP_FILTER_FLAG_TSYNC
Hi! I'm the author of [extrasafe](https://github.com/boustrophedon/extrasafe), a Rust wrapper around seccomp (and soon landlock as well, hopefully). I'm looking to switch from libseccomp-rs to seccompiler, mostly to make static compilation easier.

libseccomp supports the SECCOMP_FILTER_FLAG_TSYNC flag, which is a flag you can pass when calling the seccomp syscall directly. It allows you to apply the current seccomp filter to all running threads (TSYNC = thread sync).

Libseccomp achieves this by calling the seccomp syscall directly. It seems that [seccompiler uses prctrl](https://github.com/rust-vmm/seccompiler/blob/83dcac7071a37ed6d1a36dbe074fb3d408c615c0/src/lib.rs#L316) to enable seccomp, so in addition to adding a new flag to `seccompiler::apply_filter`, it would also need to be modified to call the syscall itself rather than using prctl.

If you'd be open to accepting a patch I'd be glad to make it - maybe just extracting the body of apply_filter into a new function `apply_filter_with_flags`, changing it to use the seccomp syscall, and then having apply_filter just proxy to `apply_filter_with_flags` with empty flags.

If you have a better design or don't want to support it at all, that's fine, just let me know!

---

Just for reference (mostly for me), here's [a convenient link](https://man7.org/linux/man-pages/man2/seccomp.2.html) to the seccomp syscall manpage

And here's the libseccomp code that calls the seccomp syscall directly, passing in the flags. See the few lines above it for where the flags are set: https://github.com/seccomp/libseccomp/blob/f1c3196d9b95de22dde8f23c5befcbeabef5711c/src/system.c#L414

### PR DESCRIPTION
**Pull request by boustrophedon**
### Summary of the PR

- Adds public function `seccompiler::apply_filter_all_threads`
- Moves the body of apply_filter into apply_filter_with_flags
- Use seccomp syscall instead of prctl to apply seccomp filters

Resolves #57 

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: src/backend/flags.rs
Comment (boustrophedon) [inline src/backend/flags.rs:1]: I can add tests to this file if necessary.

File: tests/multi_thread.rs
Comment (boustrophedon) [inline tests/multi_thread.rs:1]: This test was adapted from my tsync test in extrasafe. https://github.com/boustrophedon/extrasafe/blob/master/tests/thread_multi.rs

**Comment by boustrophedon** [comment]
Looks like musl doesn't have the linux/seccomp.h `SECCOMP_SET_MODE_FILTER` constant defined in libc? There's also lint and formatting errors that I need to fix.

File: src/backend/flags.rs
Comment (boustrophedon) [reply src/backend/flags.rs:1]: I added them so that the CI coverage tests would pass but it still decreased by 0.6%. I'm not really sure why, maybe it's counting clone and default implementations or something. It seems like it's ignoring integration_tests.rs for some reason, so maybe the new apply_filter* functions aren't getting covered.

I guess I'll try to figure out how to run them locally and either see if I can add more tests or just update the file.

**Comment by boustrophedon** [comment]
It looks like everything except coverage and musl builds are passing now. I'm not sure what the best resolution is for the missing constant in musl libc.

**Comment by rbradford** [comment]
> It looks like everything except coverage and musl builds are passing now. I'm not sure what the best resolution is for the missing constant in musl libc.

Open a PR against rust-lang/libc since this looks like it wasn't included when then gnu/musl split happened.

**Comment by boustrophedon** [comment]
https://github.com/rust-lang/libc/issues/3342

Did some digging, seems like it was just missed at some point.

In the meantime, I'll update this diff to use libc::SECCOMP_FILTER_FLAG_TSYNC inside flags.rs which I had missed before but does anyone have an issue with landing this just using a locally-defined `const SECCOMP_SET_MODE_FILTER: libc::c_int = 1` inside seccompiler until the libc PR gets merged?

**Comment by alindima** [comment]
> [rust-lang/libc#3342](https://github.com/rust-lang/libc/issues/3342)
> 
> Did some digging, seems like it was just missed at some point.
> 
> In the meantime, I'll update this diff to use libc::SECCOMP_FILTER_FLAG_TSYNC inside flags.rs which I had missed before but does anyone have an issue with landing this just using a locally-defined `const SECCOMP_SET_MODE_FILTER: libc::c_int = 1` inside seccompiler until the libc PR gets merged? It seems like that's [happening a lot in bpf.rs already](https://github.com/rust-vmm/seccompiler/blob/83dcac7071a37ed6d1a36dbe074fb3d408c615c0/src/backend/bpf.rs#L116) so maybe I can do a follow up PR after the libc PR gets merged to clean up all the constants.

Yes, this is fine until the PR is merged in rust/libc 👍🏻

File: src/backend/flags.rs
Comment (alindima) [inline src/backend/flags.rs:5]: I don't see the reason for creating such a wrapper over the libc flags.
I suggest simply accepting a `u64` in the `apply_filter_with_flags` function. The user is then responsible for setting their own flags if they so wish.

File: src/lib.rs
Comment (alindima) [inline src/lib.rs:299]: Please mention that the user should search the man page containing the `SECCOMP_FILTER_FLAG_TSYNC ` flag, as there are some important considerations there

File: CHANGELOG.md
Comment (alindima) [inline CHANGELOG.md:1]: Please mention that the underlying syscall used for installing the filter changed from `prctl` to `seccomp`.
I don't think this will be a breaking change in any way, but it's good to have in mind.

File: src/lib.rs
Comment (alindima) [inline src/lib.rs:338]: nit: move this to the top of the file. it's best to have constants defined there

File: src/lib.rs
Comment (alindima) [inline src/lib.rs:306]: Having a close look at the [man page](https://man7.org/linux/man-pages/man2/seccomp.2.html), I see that the return value of the syscall when the tsync flag is set actually is different, and contains `the ID of the thread that caused the synchronization failure`.

We should also return a new type of error wrapping this thread id.

And while we're at it, change the `Error::Prctl` variant was used with both errors from the `NO_NEW_PRIVS` call and the `SET_FILTER`. Add a new variant to the error for filter installation errors.

**Comment by boustrophedon** [comment]
Thanks for the review! I just pushed my PR to libc and am about to go to bed but I'll start working on the fixes tomorrow!

File: src/lib.rs
Comment (alindima) [inline src/lib.rs:319]: considering the various implications of adding flags to the `seccomp` syscall, I wouldn't expose it to the public.
The return type of the function can change depending on the flag (for `SECCOMP_FILTER_FLAG_NEW_LISTENER `, it returns a file descriptor).

For more advanced use cases, let's leave it to the user of the library to install the filters and handle all corner-cases themselves.

The `apply_filter` was meant as a helper for installing filters in the default use-case scenario, when no customisation was needed.

File: tests/multi_thread.rs
Comment (alindima) [inline tests/multi_thread.rs:43]: let's also try the getpid call here and see that it fails

File: tests/multi_thread.rs
Comment (alindima) [inline tests/multi_thread.rs:39]: I think this will restrict all threads in the process, including other tests that may be running and the other `cargo` threads, no?

We can spawn a process before spawning the children threads, with `libc::fork`, in order to contain the filter

**Review by alindima — COMMENTED** [review]
Thanks for the contribution! Glad to hear you're using seccompiler 😄

**Comment by boustrophedon** [comment]
I've made all the requested changes I think.

One thing I just noticed is that the manpage for seccomp says that the flags parameter is an unsigned int, but the definition for the flags in <linux/seccomp.h> is `(1UL << 0)` i.e. an unsigned long. I've changed the parameter to `libc::c_ulong` since the [TSYNC flag constant is a ulong](https://docs.rs/libc/latest/libc/constant.SECCOMP_FILTER_FLAG_TSYNC.html) and [my PR also uses ulong](https://github.com/rust-lang/libc/pull/3343).

The only thing I'm not sure about is the `source` impl for ThreadSync - currently I left it as None since i64 doesn't impl Error but maybe there's something better I could do?

File: src/lib.rs
Comment (boustrophedon) [inline src/lib.rs:255]: I'm not sure if there's something better I can do here.
`ThreadSync(pid) -> Some(pid)` doesn't work because i64 doesn't impl Error.

File: src/lib.rs
Comment (alindima) [inline src/lib.rs:373]: let's add `rc < 0` here. Otherwise, if the kernel were to return a negative value, we would return `Ok`

File: src/lib.rs
Comment (petreeftime) [inline src/lib.rs:338]: Ideally `flags` should be similar to something like OFlags in nix: https://docs.rs/nix/latest/nix/fcntl/struct.OFlag.html, it's slightly more robust, but if `apply_filter_with_flags` remain just for internal purposes (ie. `flags` is not part of the public API), then leaving it like this for now is fine.

Comment (alindima) [reply src/lib.rs:338]: This was the initial approach in the PR but since different flags for the seccomp syscall results in different return values and has different implications, I suggested to keep this function for internal use only, to keep things simple and hard to misuse

**Comment by boustrophedon** [comment]
Thanks for catching everything that I missed!

Regarding the clippy failure, I can make the change if you want but I personally think [clippy's recommendation](https://rust-lang.github.io/rust-clippy/master/index.html#/comparison_chain) is not nearly as readable as just writing "rc > 0". If I'm looking at a single arm, "Ordering::Greater" doesn't tell me which is greater - I have to refer back to the top of the match statement. `x < y` is something everyone learned how to read in elementary school.

File: tests/multi_thread.rs
Comment (alindima) [inline tests/multi_thread.rs:45]: nit: check that getpid works before installing the filter, just in case

**Review by alindima — APPROVED** [review]
Thanks

**Comment by alindima** [comment]
I agree that clippy's suggestion is less readable.
You can use this line to disable it on the function definition: `#[allow(clippy::comparison_chain)]`

**Comment by boustrophedon** [comment]
Looks like everything is passing now, thanks everyone for the reviews!

**Review by alindima — APPROVED** [review]
LGTM

**Review by petreeftime — APPROVED** [review]

**Pull request merged**
